File: Top-Interview-Questions/Strings/Medium-Collection/49-group-anagrams.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 *
 * 배열로 주어진 각 문자열들을 알파뱃 순으로 정렬하여 공통된 Anagram을 뽑아낸다.
 * Runtime: 116 ms, faster than 95.30% of JavaScript online submissions for Group Anagrams.
 * Memory Usage: 47.8 MB, less than 17.29% of JavaScript online submissions for Group Anagrams.
 * TC: O(N + KlogK), where N is the number of elements in strs and K is the average length of each string in strs
 * SC: O(N), where N is the number of elements in strs
 *
 */
var groupAnagrams = function(strs) {
    if (!strs) {
        return null;
    } else if (strs.length === 1) {
        return [strs];
    }

    const map = {};
    for (const str of strs) {
        const sortedString = sortString(str);

        if (map[sortedString]) {
            map[sortedString].push(str);
        } else {
            map[sortedString] = [str];
        }
    }

    return Object.values(map);
};

function sortString (str) {
    return Array
        .from(str)
        .sort()
        .join("");
}

File: Top-Interview-Questions/새 폴더/49-Group-Anagrams.cpp
// Time Complexity: O(size of strs * length of each string)
// Space Complexity: O(number of different anagrams * number of same anagrams)
// Runtime: 88 ms, faster than 58.16%
// Memory Usage: 22.4 MB, less than 10.55%
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
    
        vector<vector<string>> res;
        vector<int> temp1(26,0); // O(1)
        map<vector<int>, vector<int>> temp; // 
        
        for(int i = 0; i < strs.size(); i++){
            
            for(char a : strs[i]){
                temp1[a-'a']++;
            }
            temp[temp1].push_back(i);
            
            for(int j = 0; j < temp1.size(); j++){
                temp1[j] = 0;
            }
        }
        
        vector<string> t;
        
        for(auto const &it : temp){
            
            for(int i = 0; i < it.second.size(); i++){
                t.push_back(strs[it.second[i]]);
            }
            res.push_back(t);
            t.clear();
        }
        
        
        return res;
    }
};